File: backend/app/api/questions.py
"""
Questions router — full question bank CRUD.

Routes:
  GET    /api/questions              → student-safe paginated list (ExamQuestionResponse)
  GET    /api/questions/{id}         → student-safe single question (ExamQuestionResponse)
  POST   /api/questions              → create manually (QuestionResponse — admin, includes correct_answer)
  PATCH  /api/questions/{id}         → partial update (QuestionResponse — admin)
  DELETE /api/questions/{id}         → delete

Security note:
  GET endpoints use ExamQuestionResponse which NEVER includes correct_answer or explanation.
  QuestionResponse (with correct_answer) is ONLY returned from write endpoints.
"""

from __future__ import annotations

from uuid import UUID

from fastapi import APIRouter, Query, status

from app.core.dependencies import CurrentUserId, DbSession
from app.core.config import get_settings
from app.schemas.common import PaginatedResponse
from app.schemas.question import (
    Difficulty,
    ExamQuestionResponse,
    Language,
    QuestionCreateRequest,
    QuestionFilterParams,
    QuestionResponse,
    QuestionUpdateRequest,
    QuestionGenerateRequest,
    QuestionValidateBatchRequest,
    SourceType,
    VerificationStatus,
)
from app.services.question_service import QuestionService
from app.services.ai_service import AIService

router = APIRouter(prefix="/questions", tags=["Questions"])


@router.get(
    "",
    response_model=PaginatedResponse,
    summary="List questions",
    description=(
        "Returns a paginated, filterable list of questions from the question bank. "
        "Response items use ExamQuestionResponse — correct_answer and explanation are NEVER included."
    ),
)
async def list_questions(
    db: DbSession,
    _user_id: CurrentUserId,
    subject_id: UUID | None = Query(None, description="Filter by subject UUID"),
    topic_id: UUID | None = Query(None, description="Filter by topic UUID"),
    source_type: SourceType | None = Query(None, description="PYQ, AI_GENERATED, or REFERENCE"),
    difficulty: Difficulty | None = Query(None, description="EASY, MEDIUM, or HARD"),
    verification_status: VerificationStatus | None = Query(None),
    source_year: int | None = Query(None, description="Filter PYQs by exam year"),
    language: Language | None = Query(None),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
) -> PaginatedResponse:
    filters = QuestionFilterParams(
        subject_id=subject_id,
        topic_id=topic_id,
        source_type=source_type,
        difficulty=difficulty,
        verification_status=verification_status,
        source_year=source_year,
        language=language,
    )
    service = QuestionService(db)
    return await service.list_questions(filters, page, page_size)


@router.get(
    "/{question_id}",
    response_model=ExamQuestionResponse,
    summary="Get a question by ID (student-safe)",
    description=(
        "Returns question text and options. "
        "correct_answer and explanation are NEVER included in this response. "
        "Backend grading accesses correct_answer directly from the database."
    ),
)
async def get_question(
    question_id: UUID,
    db: DbSession,
    _user_id: CurrentUserId,
) -> ExamQuestionResponse:
    service = QuestionService(db)
    question = await service.get_question(question_id)
    return ExamQuestionResponse.model_validate(question)


@router.post(
    "",
    response_model=QuestionResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Create a question manually",
    description=(
        "Manually add a question to the bank. "
        "AI_GENERATED questions must have source_year=null. "
        "All manually created questions start with verification_status=UNVERIFIED."
    ),
)
async def create_question(
    body: QuestionCreateRequest,
    db: DbSession,
    _user_id: CurrentUserId,
) -> QuestionResponse:
    service = QuestionService(db)
    question = await service.create_question(body)
    return QuestionResponse.model_validate(question)


@router.patch(
    "/{question_id}",
    response_model=QuestionResponse,
    summary="Partially update a question",
    description="Update any subset of question fields. All fields are optional.",
)
async def update_question(
    question_id: UUID,
    body: QuestionUpdateRequest,
    db: DbSession,
    _user_id: CurrentUserId,
) -> QuestionResponse:
    service = QuestionService(db)
    question = await service.update_question(question_id, body)
    return QuestionResponse.model_validate(question)


@router.delete(
    "/{question_id}",
    status_code=status.HTTP_200_OK,
    summary="Delete a question",
)
async def delete_question(
    question_id: UUID,
    db: DbSession,
    _user_id: CurrentUserId,
) -> None:
    service = QuestionService(db)
    await service.delete_question(question_id)


# ── AI Endpoints ─────────────────────────────────────────────────────────────

@router.post(
    "/{question_id}/validate",
    response_model=QuestionResponse,
    summary="Validate a question using AI",
    description="Synchronously validates a single question. Updates text, options, answer, and status.",
)
async def validate_question(
    question_id: UUID,
    db: DbSession,
    _user_id: CurrentUserId,
) -> QuestionResponse:
    service = AIService(db)
    return await service.validate_question(question_id)


@router.post(
    "/validate-batch",
    status_code=status.HTTP_202_ACCEPTED,
    summary="Queue a batch of questions for validation",
    description="Asynchronously validates a batch of questions using Celery (USE_CELERY=true only).",
)
async def validate_questions_batch(
    body: QuestionValidateBatchRequest,
    db: DbSession,
    _user_id: CurrentUserId,
) -> dict:
    if not get_settings().USE_CELERY:
        from fastapi import HTTPException
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="Batch validation requires Celery. USE_CELERY is disabled in this deployment.",
        )
    from app.pdf.tasks import validate_questions_batch as task
    task.delay([str(q_id) for q_id in body.question_ids])
    return {"message": f"Queued {len(body.question_ids)} questions for validation."}


@router.post(
    "/generate",
    response_model=list[QuestionCreateRequest],
    summary="Generate new mock test questions via AI",
    description="Generates questions and returns them. Does not save them to the database.",
)
async def generate_questions(
    body: QuestionGenerateRequest,
    db: DbSession,
    _user_id: CurrentUserId,
) -> list[QuestionCreateRequest]:
    service = AIService(db)
    return await service.generate_questions(
        subject_id=body.subject_id,
        topic_id=body.topic_id,
        difficulty=body.difficulty,
        count=body.count,
    )

File: frontend/lib/api/services/questions.ts
import { fetchApi, USE_MOCKS } from '../client';
import { Question, QuestionOption } from '@/types';
import { mockQuestions } from '../mock-services';

export interface QuestionFilters {
  subjectId?: string;
  topicId?: string;
  difficulty?: string;
  sourceType?: string;
  page?: number;
  pageSize?: number;
}

export function normalizeQuestion(raw: any): Question {
  if (!raw) {
    return {
      id: '',
      subjectId: '',
      text: '',
      options: [],
      difficulty: 'MEDIUM',
    };
  }

  const id = raw.id ?? raw.question_id ?? '';
  const text = raw.text ?? raw.question_text ?? raw.question ?? '';
  const subjectId = raw.subjectId ?? raw.subject_id ?? '';
  const topicId = raw.topicId ?? raw.topic_id ?? undefined;
  const difficulty = raw.difficulty ?? 'MEDIUM';
  const sourceType = raw.sourceType ?? raw.source_type ?? 'MIXED';

  let options: QuestionOption[] = [];

  if (Array.isArray(raw.options)) {
    options = raw.options.map((opt: any, idx: number) => {
      if (typeof opt === 'string') {
        const optionId = ['A', 'B', 'C', 'D'][idx] || String(idx + 1);
        return { id: optionId, text: opt };
      }
      const optionId = opt.id ?? opt.key ?? ['A', 'B', 'C', 'D'][idx] ?? String(idx + 1);
      const optionText = opt.text ?? opt.option_text ?? opt.value ?? '';
      return { id: String(optionId), text: String(optionText) };
    });
  } else if (raw.options && typeof raw.options === 'object') {
    options = Object.entries(raw.options).map(([key, val]) => ({
      id: key,
      text: String(val),
    }));
  } else if (raw.option_a || raw.option_b || raw.option_c || raw.option_d) {
    options = [
      { id: 'A', text: String(raw.option_a || '') },
      { id: 'B', text: String(raw.option_b || '') },
      { id: 'C', text: String(raw.option_c || '') },
      { id: 'D', text: String(raw.option_d || '') },
    ].filter(o => o.text !== '');
  }

  return {
    id,
    subjectId,
    topicId,
    text,
    options,
    difficulty,
    sourceType,
    source: raw.source,
    sourceYear: raw.sourceYear ?? raw.source_year ?? null,
    sourceExam: raw.sourceExam ?? raw.source_exam ?? null,
    verificationStatus: raw.verificationStatus ?? raw.verification_status,
    pdfSourceId: raw.pdfSourceId ?? raw.pdf_source_id ?? null,
  };
}

export const questionsApi = {
  getQuestions: async (filters: QuestionFilters = {}): Promise<Question[]> => {
    if (USE_MOCKS) {
      let list = [...mockQuestions];
      if (filters.subjectId) list = list.filter(q => q.subjectId === filters.subjectId);
      if (filters.difficulty) list = list.filter(q => q.difficulty.toLowerCase() === filters.difficulty?.toLowerCase());
      return list;
    }

    try {
      const data = await fetchApi<any>('/api/questions', {
        params: {
          subjectId: filters.subjectId,
          subject_id: filters.subjectId,
          topicId: filters.topicId,
          topic_id: filters.topicId,
          difficulty: filters.difficulty?.toUpperCase(),
          source_type: filters.sourceType,
          page: filters.page,
          page_size: filters.pageSize,
        },
      });

      // Normalize if backend returns paginated object { items: Question[] } or array Question[]
      let items: any[] = [];
      if (Array.isArray(data)) {
        items = data;
      } else if (data && Array.isArray(data.items)) {
        items = data.items;
      }
      return items.map(normalizeQuestion);
    } catch (err) {
      if (USE_MOCKS) return mockQuestions;
      throw err;
    }
  },

  getById: async (id: string): Promise<Question> => {
    const raw = await fetchApi<any>(`/api/questions/${id}`);
    return normalizeQuestion(raw);
  },

  // AI Question Generation (via FastAPI backend)
  generateAIQuestions: async (payload: { subjectId: string; topicId?: string; count?: number; difficulty?: string }): Promise<Question[]> => {
    return fetchApi<Question[]>('/api/questions/generate', {
      method: 'POST',
      body: payload,
    });
  },

  // AI Question Validation (via FastAPI backend)
  validateQuestion: async (questionId: string): Promise<{ isValid: boolean; feedback?: string }> => {
    return fetchApi<{ isValid: boolean; feedback?: string }>(`/api/questions/${questionId}/validate`, {
      method: 'POST',
    });
  },
};
